{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "slnLOK7A-Sps"
   },
   "outputs": [],
   "source": [
    "from tensorflow import enable_eager_execution\n",
    "enable_eager_execution()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "5_6FWta9-V5Z"
   },
   "source": [
    "## Google Colab set up\n",
    "\n",
    "Make sure that you're using a GPU: \n",
    "\n",
    "```Runtime > Change Runtime Type > Hardware Accelerator > GPU```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 35
    },
    "colab_type": "code",
    "id": "u2qH04zm-YE0",
    "outputId": "99990a0f-f5f1-4738-82b7-7c1ca9655d25"
   },
   "outputs": [],
   "source": [
    "# import tensorflow as tf\n",
    "\n",
    "# if tf.test.gpu_device_name() != '/device:GPU:0':\n",
    "#   print('WARNING: GPU device not found.')\n",
    "# else:\n",
    "#   print('SUCCESS: Found GPU: {}'.format(tf.test.gpu_device_name()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "rjltHVVn-buD"
   },
   "source": [
    "Download code dependencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 713
    },
    "colab_type": "code",
    "id": "Nk9X4CbI-d7L",
    "outputId": "334d8b0e-fcb9-43fe-ccfa-e00e59576a6d"
   },
   "outputs": [],
   "source": [
    "# !rm load_preprocess_data.py \n",
    "# !wget https://raw.githubusercontent.com/petermchale/denoising_coverage_profiles/master/load_preprocess_data.py\n",
    "# !wget https://raw.githubusercontent.com/petermchale/denoising_coverage_profiles/master/utility.py\n",
    "# !pip install pyfaidx\n",
    "# !pip install biopython\n",
    "\n",
    "# !wget https://github.com/arq5x/bedtools2/releases/download/v2.28.0/bedtools\n",
    "# !chmod +x ./bedtools\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "UJg7xfvb-h27"
   },
   "source": [
    "Download data dependencies: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 696
    },
    "colab_type": "code",
    "id": "Smefkztr-j37",
    "outputId": "52b11ef2-9e1e-4e2d-a599-a711e5f24d89"
   },
   "outputs": [],
   "source": [
    "# !rm chr2.fa chr2.fa.gz\n",
    "# !wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr2.fa.gz\n",
    "# !gunzip chr2.fa.gz\n",
    "# # change the chromosome label in the FASTA to match that in Hao's call file\n",
    "# !sed -ie 's/chr2/2/' chr2.fa\n",
    "\n",
    "# !rm chr22.fa chr22.fa.gz\n",
    "# !wget http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr22.fa.gz\n",
    "# !gunzip chr22.fa.gz\n",
    "# !sed -ie 's/chr22/22/' chr22.fa\n",
    "\n",
    "# !wget https://gist.githubusercontent.com/petermchale/efb4140dcebf6e2a7b59760fab482b5d/raw/480c656cfe791dda4564c290f32c267860c6d7c0/download_google_drive.sh\n",
    "# !chmod +x download_google_drive.sh\n",
    "# !./download_google_drive.sh 1yJhUJYfnv_VnRKNPPFkj5BbaG-w9W-WI 100.multicov.int32.bin\n",
    "# !./download_google_drive.sh 1DIOO65uVUQU9gt6BoiVMGZZCkk_Zu-TL 1.multicov.int32.bin\n",
    "\n",
    "# !wget http://home.chpc.utah.edu/~u0875014/call.filtered.bed\n",
    "\n",
    "# !rm limo-truth.vcf\n",
    "# !wget http://home.chpc.utah.edu/~u6018199/limo-truth.vcf.gz\n",
    "# !gunzip limo-truth.vcf.gz\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "bB18Czx9SO_x"
   },
   "source": [
    "## SV calls from Hao's naive caller\n",
    "\n",
    "Hao applied his naive SV caller (of deletions only) to read-fragment-corrected depth-of-coverage data from hg002 (GRCh37; chr2; http://home.chpc.utah.edu/~u0875014/hg002.bam). \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 265
    },
    "colab_type": "code",
    "id": "je2iwlz6SO_7",
    "outputId": "8463969e-a7be-47e5-94d5-712fbd88e223"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of naive calls: 5840\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>number_of_Ns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10183</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>9999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10890</td>\n",
       "      <td>11120</td>\n",
       "      <td>ggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagag...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>133229</td>\n",
       "      <td>133339</td>\n",
       "      <td>aaaattatcctggcatggtggcgtatgcctgtagtcccaaccgcacaggaggcttaggtgggaggatccactgaactcaggaggtggtggatgcag...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>197732</td>\n",
       "      <td>198231</td>\n",
       "      <td>GCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCGTTTTCCCTTCTGTCTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>198238</td>\n",
       "      <td>198395</td>\n",
       "      <td>TGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome   start     end  \\\n",
       "0          2       1   10183   \n",
       "1          2   10890   11120   \n",
       "2          2  133229  133339   \n",
       "3          2  197732  198231   \n",
       "4          2  198238  198395   \n",
       "\n",
       "                                                                                              sequence  \\\n",
       "0  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...   \n",
       "1  ggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagag...   \n",
       "2  aaaattatcctggcatggtggcgtatgcctgtagtcccaaccgcacaggaggcttaggtgggaggatccactgaactcaggaggtggtggatgcag...   \n",
       "3  GCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCGTTTTCCCTTCTGTCTG...   \n",
       "4  TGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTT...   \n",
       "\n",
       "   number_of_Ns  \n",
       "0          9999  \n",
       "1             0  \n",
       "2             0  \n",
       "3             0  \n",
       "4             0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import OrderedDict    \n",
    "import pandas as pd\n",
    "pd.set_option('max_colwidth', 100)\n",
    "import numpy as np \n",
    "\n",
    "def calls_to_dataframe(calls):\n",
    "    chromosomes, starts, ends, sequences, number_of_Ns = [], [], [], [], []\n",
    "    for call in calls: \n",
    "        chromosome, start, end, sequence = call.replace('-', '\\t').replace(':', '\\t').split('\\t')\n",
    "        chromosomes.append(chromosome)\n",
    "        starts.append(start)\n",
    "        ends.append(end)\n",
    "        sequences.append(sequence)\n",
    "        number_of_Ns.append(sequence.upper().count('N'))\n",
    "        \n",
    "    return pd.DataFrame(OrderedDict([\n",
    "        ('chromosome', chromosomes),\n",
    "        ('start', starts),\n",
    "        ('end', ends),\n",
    "        ('sequence', sequences),\n",
    "        ('number_of_Ns', number_of_Ns)\n",
    "    ]))    \n",
    "\n",
    "def path(file): \n",
    "#     return './' + file\n",
    "    return '../data/temp/' + file\n",
    "\n",
    "def get_naive_calls(): \n",
    "    call_set = path('call.filtered.bed')\n",
    "    chr2 = path('chr2.fa')\n",
    "    calls = !bedtools getfasta -bed $call_set -fi $chr2 -fo /dev/stdout -tab 2>info \n",
    "    return calls_to_dataframe(calls)\n",
    "\n",
    "naive_calls = get_naive_calls()\n",
    "print('number of naive calls: {}'.format(len(naive_calls)))\n",
    "naive_calls.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "tXjRV_1K58vl"
   },
   "source": [
    "## False calls from naive SV caller\n",
    "\n",
    "Calls that are not present in a truth set for this sample are deemed \"false calls\". \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 239
    },
    "colab_type": "code",
    "id": "hQt0_TsN58vl",
    "outputId": "27962df2-581a-4392-c2b9-27b8fbd31d37"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of false naive calls: 4369\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>number_of_Ns</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10183</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>9999</td>\n",
       "      <td>10182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10890</td>\n",
       "      <td>11120</td>\n",
       "      <td>ggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagag...</td>\n",
       "      <td>0</td>\n",
       "      <td>230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>133229</td>\n",
       "      <td>133339</td>\n",
       "      <td>aaaattatcctggcatggtggcgtatgcctgtagtcccaaccgcacaggaggcttaggtgggaggatccactgaactcaggaggtggtggatgcag...</td>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>197732</td>\n",
       "      <td>198231</td>\n",
       "      <td>GCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCGTTTTCCCTTCTGTCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>198238</td>\n",
       "      <td>198395</td>\n",
       "      <td>TGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTT...</td>\n",
       "      <td>0</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome   start     end  \\\n",
       "0          2       1   10183   \n",
       "1          2   10890   11120   \n",
       "2          2  133229  133339   \n",
       "3          2  197732  198231   \n",
       "4          2  198238  198395   \n",
       "\n",
       "                                                                                              sequence  \\\n",
       "0  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...   \n",
       "1  ggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagagaggcgcaccgcgcccgcgcaggcgcagag...   \n",
       "2  aaaattatcctggcatggtggcgtatgcctgtagtcccaaccgcacaggaggcttaggtgggaggatccactgaactcaggaggtggtggatgcag...   \n",
       "3  GCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCGTTTTCCCTTCTGTCTG...   \n",
       "4  TGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTTTTCCCTTCTGTCTGCCTTTCCGGCCTGTTGACGTGGAGACGCGCGGGGCCTT...   \n",
       "\n",
       "   number_of_Ns  length  \n",
       "0          9999   10182  \n",
       "1             0     230  \n",
       "2             0     110  \n",
       "3             0     499  \n",
       "4             0     157  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_false_naive_calls(): \n",
    "    call_set = path('call.filtered.bed')\n",
    "    false_call_set = path('call.filtered.false.bed')\n",
    "    truth_set = path('limo-truth.vcf')\n",
    "    chr2 = path('chr2.fa')\n",
    "    !bedtools intersect -v -a $call_set -b $truth_set > $false_call_set 2>> info\n",
    "    calls = !bedtools getfasta -bed $false_call_set -fi $chr2 -fo /dev/stdout -tab 2>> info\n",
    "    df = calls_to_dataframe(calls)\n",
    "    df['length'] = df['sequence'].map(lambda s: len(s))\n",
    "    return df\n",
    "\n",
    "false_naive_calls = get_false_naive_calls()\n",
    "print('number of false naive calls: {}'.format(len(false_naive_calls)))\n",
    "false_naive_calls.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "GDR7LABB58vv"
   },
   "source": [
    "## False SV calls containing AT repeats\n",
    "\n",
    "[Ross et al 2013](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-5-r51) (Table 2), [Smith et al 2015](https://peerj.com/articles/836/) (Fig 3), and we (CEPH data) have established that sequences containing AT dinucleotide repeats tend to have lower read depths than random sequences. This systematic reduction in depth could have fooled Hao's caller into making a call where no SV exists.  \n",
    "\n",
    "First, let's get false calls containing AT repeats: \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "Vq-d7YFS58vw"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>number_of_Ns</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>2</td>\n",
       "      <td>87778896</td>\n",
       "      <td>87778998</td>\n",
       "      <td>ATTTTGGCCTGAGTTCTGCCTTGGATTTATGTTTTTTGTtacatatatatatatatatatatatatattttttttttttttttttttttttttttt...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1779</th>\n",
       "      <td>2</td>\n",
       "      <td>92206632</td>\n",
       "      <td>92222959</td>\n",
       "      <td>ACTCTTAATGATACAACAGCTAAATATAGGTCTAATGCTCATTCCGTGTGGACAACAATAGCAGCCATTCCCACAAATGGCTGATTTGTGGGAAGT...</td>\n",
       "      <td>0</td>\n",
       "      <td>16327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2432</th>\n",
       "      <td>2</td>\n",
       "      <td>110746778</td>\n",
       "      <td>110746954</td>\n",
       "      <td>gcattgtttaaatgagttataagattgagcatgttttcatgtttattggctgcttgtatatcttctttggagaaatgtctattcagattgtctccc...</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>2</td>\n",
       "      <td>110746930</td>\n",
       "      <td>110747194</td>\n",
       "      <td>gtatatatatatatatatatatatttttttttttttttttttttgagacggagtcttactctgttgccaggctggagtgcagtggcacgatcttgg...</td>\n",
       "      <td>0</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3359</th>\n",
       "      <td>2</td>\n",
       "      <td>131299237</td>\n",
       "      <td>131305361</td>\n",
       "      <td>GCCATATCTGACCCTTTCTCTCTGTCTAAGGAAATGGGGCAAGGCAGGGGAAGCAGGCCCAGTGGTGCACAGTGGGCTGTTGGTGTCAGGAGGCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>6124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3361</th>\n",
       "      <td>2</td>\n",
       "      <td>131306037</td>\n",
       "      <td>131316741</td>\n",
       "      <td>aattcctggactcatacaatcctcccaccttagcctcccaagtactaggattacaggtgtgagccatcacacccggcTCCATTTAGCTATTTCAAA...</td>\n",
       "      <td>0</td>\n",
       "      <td>10704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3796</th>\n",
       "      <td>2</td>\n",
       "      <td>175459741</td>\n",
       "      <td>175459872</td>\n",
       "      <td>ccagggtgggaagatagcttgagcccaagagtttgagaccatcctgggcaaacatagtgggaccctcatctctacaaaaaaaaaaaaaaaaaaaaa...</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome      start        end  \\\n",
       "1022          2   87778896   87778998   \n",
       "1779          2   92206632   92222959   \n",
       "2432          2  110746778  110746954   \n",
       "2433          2  110746930  110747194   \n",
       "3359          2  131299237  131305361   \n",
       "3361          2  131306037  131316741   \n",
       "3796          2  175459741  175459872   \n",
       "\n",
       "                                                                                                 sequence  \\\n",
       "1022  ATTTTGGCCTGAGTTCTGCCTTGGATTTATGTTTTTTGTtacatatatatatatatatatatatatattttttttttttttttttttttttttttt...   \n",
       "1779  ACTCTTAATGATACAACAGCTAAATATAGGTCTAATGCTCATTCCGTGTGGACAACAATAGCAGCCATTCCCACAAATGGCTGATTTGTGGGAAGT...   \n",
       "2432  gcattgtttaaatgagttataagattgagcatgttttcatgtttattggctgcttgtatatcttctttggagaaatgtctattcagattgtctccc...   \n",
       "2433  gtatatatatatatatatatatatttttttttttttttttttttgagacggagtcttactctgttgccaggctggagtgcagtggcacgatcttgg...   \n",
       "3359  GCCATATCTGACCCTTTCTCTCTGTCTAAGGAAATGGGGCAAGGCAGGGGAAGCAGGCCCAGTGGTGCACAGTGGGCTGTTGGTGTCAGGAGGCTG...   \n",
       "3361  aattcctggactcatacaatcctcccaccttagcctcccaagtactaggattacaggtgtgagccatcacacccggcTCCATTTAGCTATTTCAAA...   \n",
       "3796  ccagggtgggaagatagcttgagcccaagagtttgagaccatcctgggcaaacatagtgggaccctcatctctacaaaaaaaaaaaaaaaaaaaaa...   \n",
       "\n",
       "      number_of_Ns  length  \n",
       "1022             0     102  \n",
       "1779             0   16327  \n",
       "2432             0     176  \n",
       "2433             0     264  \n",
       "3359             0    6124  \n",
       "3361             0   10704  \n",
       "3796             0     131  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def sequences_containing_motif(df, motif_):\n",
    "    motif_in_sequence = df['sequence'].map(lambda sequence: motif_.upper() in sequence.upper())\n",
    "    return df[motif_in_sequence]\n",
    "\n",
    "false_naive_calls_containing_AT = sequences_containing_motif(false_naive_calls, 'AT'*10)\n",
    "false_naive_calls_containing_AT\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, let's compute the depths along each call:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>number_of_Ns</th>\n",
       "      <th>length</th>\n",
       "      <th>observed_depths</th>\n",
       "      <th>uncorrected_depths</th>\n",
       "      <th>mean_uncorrected_depths</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>2</td>\n",
       "      <td>87778896</td>\n",
       "      <td>87778998</td>\n",
       "      <td>ATTTTGGCCTGAGTTCTGCCTTGGATTTATGTTTTTTGTtacatatatatatatatatatatatatattttttttttttttttttttttttttttt...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>[14, 14, 14, 14, 14, 13, 13, 13, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 11, 11, 11, 10, 10,...</td>\n",
       "      <td>[0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, ...</td>\n",
       "      <td>0.391232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1779</th>\n",
       "      <td>2</td>\n",
       "      <td>92206632</td>\n",
       "      <td>92222959</td>\n",
       "      <td>ACTCTTAATGATACAACAGCTAAATATAGGTCTAATGCTCATTCCGTGTGGACAACAATAGCAGCCATTCCCACAAATGGCTGATTTGTGGGAAGT...</td>\n",
       "      <td>0</td>\n",
       "      <td>16327</td>\n",
       "      <td>[31, 32, 32, 32, 32, 32, 32, 32, 31, 31, 32, 31, 31, 30, 30, 30, 30, 29, 29, 28, 29, 29, 28, 29,...</td>\n",
       "      <td>[1.218793127131533, 1.2581090344583568, 1.2581090344583568, 1.2581090344583568, 1.25810903445835...</td>\n",
       "      <td>1.191399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2432</th>\n",
       "      <td>2</td>\n",
       "      <td>110746778</td>\n",
       "      <td>110746954</td>\n",
       "      <td>gcattgtttaaatgagttataagattgagcatgttttcatgtttattggctgcttgtatatcttctttggagaaatgtctattcagattgtctccc...</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>[27, 28, 28, 29, 31, 31, 31, 31, 31, 30, 30, 30, 30, 30, 30, 30, 28, 27, 27, 27, 26, 26, 26, 27,...</td>\n",
       "      <td>[1.0615294978242384, 1.100845405151062, 1.100845405151062, 1.1401613124778858, 1.218793127131533...</td>\n",
       "      <td>0.626374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>2</td>\n",
       "      <td>110746930</td>\n",
       "      <td>110747194</td>\n",
       "      <td>gtatatatatatatatatatatatttttttttttttttttttttgagacggagtcttactctgttgccaggctggagtgcagtggcacgatcttgg...</td>\n",
       "      <td>0</td>\n",
       "      <td>264</td>\n",
       "      <td>[3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 4,...</td>\n",
       "      <td>[0.11794772198047093, 0.11794772198047093, 0.11794772198047093, 0.0786318146536473, 0.0786318146...</td>\n",
       "      <td>0.633671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3359</th>\n",
       "      <td>2</td>\n",
       "      <td>131299237</td>\n",
       "      <td>131305361</td>\n",
       "      <td>GCCATATCTGACCCTTTCTCTCTGTCTAAGGAAATGGGGCAAGGCAGGGGAAGCAGGCCCAGTGGTGCACAGTGGGCTGTTGGTGTCAGGAGGCTG...</td>\n",
       "      <td>0</td>\n",
       "      <td>6124</td>\n",
       "      <td>[12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 12, 12, 12, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13,...</td>\n",
       "      <td>[0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.471790887...</td>\n",
       "      <td>0.597128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3361</th>\n",
       "      <td>2</td>\n",
       "      <td>131306037</td>\n",
       "      <td>131316741</td>\n",
       "      <td>aattcctggactcatacaatcctcccaccttagcctcccaagtactaggattacaggtgtgagccatcacacccggcTCCATTTAGCTATTTCAAA...</td>\n",
       "      <td>0</td>\n",
       "      <td>10704</td>\n",
       "      <td>[15, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 19, 18, 18, 18, 17, 16, 16, 16, 16, 16, 16, 16, 16,...</td>\n",
       "      <td>[0.5897386099023547, 0.5897386099023547, 0.5897386099023547, 0.6290545172291784, 0.6290545172291...</td>\n",
       "      <td>0.818255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3796</th>\n",
       "      <td>2</td>\n",
       "      <td>175459741</td>\n",
       "      <td>175459872</td>\n",
       "      <td>ccagggtgggaagatagcttgagcccaagagtttgagaccatcctgggcaaacatagtgggaccctcatctctacaaaaaaaaaaaaaaaaaaaaa...</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>[11, 11, 11, 10, 10, 10, 10, 10, 10, 10, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9...</td>\n",
       "      <td>[0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.39315907326823646, 0.393159073268...</td>\n",
       "      <td>0.282714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome      start        end  \\\n",
       "1022          2   87778896   87778998   \n",
       "1779          2   92206632   92222959   \n",
       "2432          2  110746778  110746954   \n",
       "2433          2  110746930  110747194   \n",
       "3359          2  131299237  131305361   \n",
       "3361          2  131306037  131316741   \n",
       "3796          2  175459741  175459872   \n",
       "\n",
       "                                                                                                 sequence  \\\n",
       "1022  ATTTTGGCCTGAGTTCTGCCTTGGATTTATGTTTTTTGTtacatatatatatatatatatatatatattttttttttttttttttttttttttttt...   \n",
       "1779  ACTCTTAATGATACAACAGCTAAATATAGGTCTAATGCTCATTCCGTGTGGACAACAATAGCAGCCATTCCCACAAATGGCTGATTTGTGGGAAGT...   \n",
       "2432  gcattgtttaaatgagttataagattgagcatgttttcatgtttattggctgcttgtatatcttctttggagaaatgtctattcagattgtctccc...   \n",
       "2433  gtatatatatatatatatatatatttttttttttttttttttttgagacggagtcttactctgttgccaggctggagtgcagtggcacgatcttgg...   \n",
       "3359  GCCATATCTGACCCTTTCTCTCTGTCTAAGGAAATGGGGCAAGGCAGGGGAAGCAGGCCCAGTGGTGCACAGTGGGCTGTTGGTGTCAGGAGGCTG...   \n",
       "3361  aattcctggactcatacaatcctcccaccttagcctcccaagtactaggattacaggtgtgagccatcacacccggcTCCATTTAGCTATTTCAAA...   \n",
       "3796  ccagggtgggaagatagcttgagcccaagagtttgagaccatcctgggcaaacatagtgggaccctcatctctacaaaaaaaaaaaaaaaaaaaaa...   \n",
       "\n",
       "      number_of_Ns  length  \\\n",
       "1022             0     102   \n",
       "1779             0   16327   \n",
       "2432             0     176   \n",
       "2433             0     264   \n",
       "3359             0    6124   \n",
       "3361             0   10704   \n",
       "3796             0     131   \n",
       "\n",
       "                                                                                          observed_depths  \\\n",
       "1022  [14, 14, 14, 14, 14, 13, 13, 13, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 11, 11, 11, 10, 10,...   \n",
       "1779  [31, 32, 32, 32, 32, 32, 32, 32, 31, 31, 32, 31, 31, 30, 30, 30, 30, 29, 29, 28, 29, 29, 28, 29,...   \n",
       "2432  [27, 28, 28, 29, 31, 31, 31, 31, 31, 30, 30, 30, 30, 30, 30, 30, 28, 27, 27, 27, 26, 26, 26, 27,...   \n",
       "2433  [3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 4,...   \n",
       "3359  [12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 12, 12, 12, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13,...   \n",
       "3361  [15, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 19, 18, 18, 18, 17, 16, 16, 16, 16, 16, 16, 16, 16,...   \n",
       "3796  [11, 11, 11, 10, 10, 10, 10, 10, 10, 10, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9...   \n",
       "\n",
       "                                                                                       uncorrected_depths  \\\n",
       "1022  [0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, ...   \n",
       "1779  [1.218793127131533, 1.2581090344583568, 1.2581090344583568, 1.2581090344583568, 1.25810903445835...   \n",
       "2432  [1.0615294978242384, 1.100845405151062, 1.100845405151062, 1.1401613124778858, 1.218793127131533...   \n",
       "2433  [0.11794772198047093, 0.11794772198047093, 0.11794772198047093, 0.0786318146536473, 0.0786318146...   \n",
       "3359  [0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.471790887...   \n",
       "3361  [0.5897386099023547, 0.5897386099023547, 0.5897386099023547, 0.6290545172291784, 0.6290545172291...   \n",
       "3796  [0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.39315907326823646, 0.393159073268...   \n",
       "\n",
       "      mean_uncorrected_depths  \n",
       "1022                 0.391232  \n",
       "1779                 1.191399  \n",
       "2432                 0.626374  \n",
       "2433                 0.633671  \n",
       "3359                 0.597128  \n",
       "3361                 0.818255  \n",
       "3796                 0.282714  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from utility import named_tuple\n",
    "import os \n",
    "\n",
    "def fetch_chromosome(df):\n",
    "    return int(df['chromosome'].iloc[0])\n",
    "\n",
    "def read_depths_new(args):\n",
    "    sample, chromosome, per_base, tool, dtype, suffix = os.path.basename(args.depth_file_name).split('.')\n",
    "    assert (chromosome == 'chr' + str(args.target_chromosome_number))\n",
    "    assert (per_base == 'per-base')\n",
    "    assert (tool == args.target_tool)\n",
    "    assert (suffix == 'bin')\n",
    "    return np.fromfile(args.depth_file_name, dtype=getattr(np, dtype))\n",
    "\n",
    "def add_depths(df, depth_file_name): \n",
    "    args = named_tuple({\n",
    "        'depth_file_name': depth_file_name,\n",
    "        'target_chromosome_number': fetch_chromosome(df), \n",
    "        'target_tool': 'mosdepth'\n",
    "    })\n",
    "    depths = read_depths_new(args)\n",
    "\n",
    "    def depths_and_meanDepth(row): \n",
    "        start = np.int32(row['start'])\n",
    "        end = np.int32(row['end'])\n",
    "        depths_slice = depths[start:end]\n",
    "        uncorrected_depths = depths_slice/np.mean(depths)\n",
    "        mean_uncorrected_depths = np.mean(uncorrected_depths)\n",
    "        return pd.Series(\n",
    "            [depths_slice, uncorrected_depths, mean_uncorrected_depths], \n",
    "            index=['observed_depths', 'uncorrected_depths', 'mean_uncorrected_depths']\n",
    "        )\n",
    "    \n",
    "    series = df.apply(depths_and_meanDepth, axis=1)\n",
    "    return pd.concat([df, series], axis=1)\n",
    "    \n",
    "false_naive_calls_containing_AT = add_depths(\n",
    "    false_naive_calls_containing_AT, \n",
    "    depth_file_name='../data/temp/hg002.chr2.per-base.mosdepth.int32.bin'\n",
    ")\n",
    "false_naive_calls_containing_AT\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, let's plot the depth profile across each call (in order of increasing effects on depth):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "linkText": "Export to plot.ly",
        "plotlyServerURL": "https://plot.ly",
        "responsive": true,
        "scrollZoom": true,
        "showLink": false
       },
       "data": [
        {
         "type": "bar",
         "uid": "dc7a50a5-5f03-4d4c-bd1c-1e8857673d48",
         "x": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
          29,
          30,
          31,
          32,
          33,
          34,
          35,
          36,
          37,
          38,
          39,
          40,
          41,
          42,
          43,
          44,
          45,
          46,
          47,
          48,
          49,
          50,
          51,
          52,
          53,
          54,
          55,
          56,
          57,
          58,
          59,
          60,
          61,
          62,
          63,
          64,
          65,
          66,
          67,
          68,
          69,
          70,
          71,
          72,
          73,
          74,
          75,
          76,
          77,
          78,
          79,
          80,
          81,
          82,
          83,
          84,
          85,
          86,
          87,
          88,
          89,
          90,
          91,
          92,
          93,
          94,
          95,
          96,
          97,
          98,
          99,
          100,
          101
         ],
         "y": [
          0.550422702575531,
          0.550422702575531,
          0.550422702575531,
          0.550422702575531,
          0.550422702575531,
          0.5111067952487074,
          0.5111067952487074,
          0.5111067952487074,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.39315907326823646,
          0.39315907326823646,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.39315907326823646,
          0.39315907326823646,
          0.39315907326823646,
          0.3538431659414128,
          0.3538431659414128,
          0.3538431659414128,
          0.3538431659414128,
          0.3538431659414128,
          0.39315907326823646,
          0.4324749805950601,
          0.4324749805950601,
          0.3538431659414128,
          0.3538431659414128,
          0.3538431659414128,
          0.3538431659414128,
          0.3145272586145892,
          0.3145272586145892,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.23589544396094186,
          0.23589544396094186,
          0.23589544396094186,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.23589544396094186,
          0.19657953663411823,
          0.2752113512877655,
          0.23589544396094186,
          0.23589544396094186,
          0.23589544396094186,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.2752113512877655,
          0.3145272586145892,
          0.3145272586145892,
          0.3145272586145892,
          0.3538431659414128,
          0.3538431659414128,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.4324749805950601,
          0.39315907326823646,
          0.39315907326823646,
          0.39315907326823646,
          0.39315907326823646,
          0.39315907326823646,
          0.39315907326823646,
          0.4324749805950601,
          0.47179088792188373,
          0.47179088792188373,
          0.47179088792188373,
          0.550422702575531,
          0.550422702575531,
          0.550422702575531,
          0.550422702575531,
          0.5897386099023547
         ]
        }
       ],
       "layout": {
        "xaxis": {
         "tickangle": 0,
         "ticktext": [
          "A",
          "T",
          "T",
          "T",
          "T",
          "G",
          "G",
          "C",
          "C",
          "T",
          "G",
          "A",
          "G",
          "T",
          "T",
          "C",
          "T",
          "G",
          "C",
          "C",
          "T",
          "T",
          "G",
          "G",
          "A",
          "T",
          "T",
          "T",
          "A",
          "T",
          "G",
          "T",
          "T",
          "T",
          "T",
          "T",
          "T",
          "G",
          "T",
          "t",
          "a",
          "c",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "a",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "t",
          "g",
          "a"
         ],
         "tickvals": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
          29,
          30,
          31,
          32,
          33,
          34,
          35,
          36,
          37,
          38,
          39,
          40,
          41,
          42,
          43,
          44,
          45,
          46,
          47,
          48,
          49,
          50,
          51,
          52,
          53,
          54,
          55,
          56,
          57,
          58,
          59,
          60,
          61,
          62,
          63,
          64,
          65,
          66,
          67,
          68,
          69,
          70,
          71,
          72,
          73,
          74,
          75,
          76,
          77,
          78,
          79,
          80,
          81,
          82,
          83,
          84,
          85,
          86,
          87,
          88,
          89,
          90,
          91,
          92,
          93,
          94,
          95,
          96,
          97,
          98,
          99,
          100,
          101
         ]
        },
        "yaxis": {
         "fixedrange": true
        }
       }
      },
      "text/html": [
       "<div>\n",
       "        \n",
       "        \n",
       "            <div id=\"7bf23611-8b0b-4950-8057-286da42949d2\" class=\"plotly-graph-div\" style=\"height:525px; width:100%;\"></div>\n",
       "            <script type=\"text/javascript\">\n",
       "                require([\"plotly\"], function(Plotly) {\n",
       "                    window.PLOTLYENV=window.PLOTLYENV || {};\n",
       "                    window.PLOTLYENV.BASE_URL='https://plot.ly';\n",
       "                    \n",
       "                if (document.getElementById(\"7bf23611-8b0b-4950-8057-286da42949d2\")) {\n",
       "                    Plotly.newPlot(\n",
       "                        '7bf23611-8b0b-4950-8057-286da42949d2',\n",
       "                        [{\"type\": \"bar\", \"uid\": \"447912b4-ac80-4495-8f65-23c5f3112447\", \"x\": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101], \"y\": [0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.5111067952487074, 0.5111067952487074, 0.5111067952487074, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.39315907326823646, 0.39315907326823646, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.39315907326823646, 0.39315907326823646, 0.39315907326823646, 0.3538431659414128, 0.3538431659414128, 0.3538431659414128, 0.3538431659414128, 0.3538431659414128, 0.39315907326823646, 0.4324749805950601, 0.4324749805950601, 0.3538431659414128, 0.3538431659414128, 0.3538431659414128, 0.3538431659414128, 0.3145272586145892, 0.3145272586145892, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.23589544396094186, 0.23589544396094186, 0.23589544396094186, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.23589544396094186, 0.19657953663411823, 0.2752113512877655, 0.23589544396094186, 0.23589544396094186, 0.23589544396094186, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.2752113512877655, 0.3145272586145892, 0.3145272586145892, 0.3145272586145892, 0.3538431659414128, 0.3538431659414128, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.4324749805950601, 0.39315907326823646, 0.39315907326823646, 0.39315907326823646, 0.39315907326823646, 0.39315907326823646, 0.39315907326823646, 0.4324749805950601, 0.47179088792188373, 0.47179088792188373, 0.47179088792188373, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.550422702575531, 0.5897386099023547]}],\n",
       "                        {\"xaxis\": {\"tickangle\": 0, \"ticktext\": [\"A\", \"T\", \"T\", \"T\", \"T\", \"G\", \"G\", \"C\", \"C\", \"T\", \"G\", \"A\", \"G\", \"T\", \"T\", \"C\", \"T\", \"G\", \"C\", \"C\", \"T\", \"T\", \"G\", \"G\", \"A\", \"T\", \"T\", \"T\", \"A\", \"T\", \"G\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"G\", \"T\", \"t\", \"a\", \"c\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"a\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"t\", \"g\", \"a\"], \"tickvals\": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101]}, \"yaxis\": {\"fixedrange\": true}},\n",
       "                        {\"scrollZoom\": true, \"plotlyServerURL\": \"https://plot.ly\", \"linkText\": \"Export to plot.ly\", \"showLink\": false, \"responsive\": true}\n",
       "                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('7bf23611-8b0b-4950-8057-286da42949d2');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })\n",
       "                };\n",
       "                });\n",
       "            </script>\n",
       "        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "nucleotides = list(false_naive_calls_containing_AT.iloc[false_naive_call_containing_AT]['sequence'])\n",
    "uncorrected_depths = list(false_naive_calls_containing_AT.iloc[false_naive_call_containing_AT]['uncorrected_depths'])\n",
    "number_nucleotides = len(nucleotides)\n",
    "\n",
    "data = [go.Bar(\n",
    "    x=np.arange(number_nucleotides),\n",
    "    y=uncorrected_depths\n",
    ")]\n",
    "\n",
    "layout = go.Layout(\n",
    "    yaxis=go.layout.YAxis(\n",
    "        fixedrange=True\n",
    "    ),\n",
    "    xaxis=go.layout.XAxis(\n",
    "        ticktext=x,\n",
    "        tickvals=np.arange(len(y)),\n",
    "        tickangle=0\n",
    "    )\n",
    ")\n",
    "\n",
    "config={'scrollZoom': True}\n",
    "\n",
    "fig = go.Figure(data, layout)\n",
    "\n",
    "plotly.offline.init_notebook_mode(connected=True)\n",
    "\n",
    "plotly.offline.iplot(fig, config=config)\n",
    "\n",
    "# py.iplot(data, filename='jupyter-basic_bar')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "E1GuFvio58wI"
   },
   "source": [
    "## To Do \n",
    "\n",
    "* make calls on more samples so that we can restrict false SV calls to those that recur in multiple individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "_Kb5gxrLSPAe"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "denoise_false_SVs_containing_ATrepeat.ipynb",
   "provenance": [],
   "toc_visible": true,
   "version": "0.3.2"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
